File: Sources/XMTPiOS/PrivatePreferences.swift
import Foundation
import LibXMTP

public enum ConsentState: String, Codable {
	case allowed, denied, unknown
}
public enum EntryType: String, Codable {
	case address, conversation_id, inbox_id
}
public enum PreferenceType: String, Codable {
	case hmac_keys
}

public struct ConsentRecord: Codable, Hashable {
	public init(value: String, entryType: EntryType, consentType: ConsentState)
	{
		self.value = value
		self.entryType = entryType
		self.consentType = consentType
	}

	static func address(_ address: String, type: ConsentState = .unknown)
		-> ConsentRecord
	{
		ConsentRecord(value: address, entryType: .address, consentType: type)
	}

	static func conversationId(
		conversationId: String, type: ConsentState = ConsentState.unknown
	) -> ConsentRecord {
		ConsentRecord(
			value: conversationId, entryType: .conversation_id,
			consentType: type)
	}

	static func inboxId(_ inboxId: String, type: ConsentState = .unknown)
		-> ConsentRecord
	{
		ConsentRecord(
			value: inboxId, entryType: .inbox_id, consentType: type)
	}

	public var value: String
	public var entryType: EntryType
	public var consentType: ConsentState

	var key: String {
		"\(entryType)-\(value)"
	}
}

/// Provides access to contact bundles.
public actor PrivatePreferences {
	var client: Client
	var ffiClient: FfiXmtpClient

	init(client: Client, ffiClient: FfiXmtpClient) {
		self.client = client
		self.ffiClient = ffiClient
	}

	public func setConsentState(entries: [ConsentRecord]) async throws {
		try await ffiClient.setConsentStates(records: entries.map(\.toFFI))
	}

	public func addressState(address: String) async throws -> ConsentState {
		return try await ffiClient.getConsentState(
			entityType: .address,
			entity: address
		).fromFFI
	}

	public func conversationState(conversationId: String) async throws
		-> ConsentState
	{
		return try await ffiClient.getConsentState(
			entityType: .conversationId,
			entity: conversationId
		).fromFFI
	}

	public func inboxIdState(inboxId: String) async throws -> ConsentState {
		return try await ffiClient.getConsentState(
			entityType: .inboxId,
			entity: inboxId
		).fromFFI
	}

	public func syncConsent() async throws {
		try await ffiClient.sendSyncRequest(kind: .consent)
	}

	public func streamConsent()
		-> AsyncThrowingStream<ConsentRecord, Error>
	{
		AsyncThrowingStream { continuation in
			let ffiStreamActor = FfiStreamActor()

			let consentCallback = ConsentCallback(client: self.client) {
				records in
				guard !Task.isCancelled else {
					continuation.finish()
					Task {
						await ffiStreamActor.endStream()
					}
					return
				}
				for consent in records {
					continuation.yield(consent.fromFfi)
				}
			}

			let task = Task {
				let stream = await ffiClient.conversations().streamConsent(
					callback: consentCallback)
				await ffiStreamActor.setFfiStream(stream)
			}

			continuation.onTermination = { _ in
				task.cancel()
				Task {
					await ffiStreamActor.endStream()
				}
			}
		}
	}

	public func streamPreferenceUpdates()
		-> AsyncThrowingStream<PreferenceType, Error>
	{
		AsyncThrowingStream { continuation in
			let ffiStreamActor = FfiStreamActor()

			let preferenceCallback = PreferenceCallback(client: self.client) {
				records in
				guard !Task.isCancelled else {
					continuation.finish()
					Task {
						await ffiStreamActor.endStream()
					}
					return
				}
				for preference in records {
					if case .hmac(let key) = preference {
						continuation.yield(.hmac_keys)
					}
				}
			}

			let task = Task {
				let stream = await ffiClient.conversations().streamPreferences(
					callback: preferenceCallback)
				await ffiStreamActor.setFfiStream(stream)
			}

			continuation.onTermination = { _ in
				task.cancel()
				Task {
					await ffiStreamActor.endStream()
				}
			}
		}
	}
}

final class ConsentCallback: FfiConsentCallback {
	let client: Client
	let callback: ([FfiConsent]) -> Void

	init(client: Client, _ callback: @escaping ([FfiConsent]) -> Void) {
		self.client = client
		self.callback = callback
	}

	func onConsentUpdate(consent: [FfiConsent]) {
		callback(consent)
	}

	func onError(error: FfiSubscribeError) {
		print("Error ConsentCallback \(error)")
	}
}

final class PreferenceCallback: FfiPreferenceCallback {
	let client: Client
	let callback: ([FfiPreferenceUpdate]) -> Void

	init(client: Client, _ callback: @escaping ([FfiPreferenceUpdate]) -> Void)
	{
		self.client = client
		self.callback = callback
	}

	func onPreferenceUpdate(preference: [FfiPreferenceUpdate]) {
		callback(preference)
	}

	func onError(error: FfiSubscribeError) {
		print("Error ConsentCallback \(error)")
	}
}

File: sdks/android/library/src/main/java/org/xmtp/android/library/PrivatePreferences.kt
package org.xmtp.android.library

import android.util.Log
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import uniffi.xmtpv3.FfiConsent
import uniffi.xmtpv3.FfiConsentCallback
import uniffi.xmtpv3.FfiConsentEntityType
import uniffi.xmtpv3.FfiConsentState
import uniffi.xmtpv3.FfiException
import uniffi.xmtpv3.FfiPreferenceCallback
import uniffi.xmtpv3.FfiPreferenceUpdate
import uniffi.xmtpv3.FfiXmtpClient

enum class ConsentState {
    ALLOWED,
    DENIED,
    UNKNOWN,
    ;

    companion object {
        fun toFfiConsentState(option: ConsentState): FfiConsentState =
            when (option) {
                ALLOWED -> FfiConsentState.ALLOWED
                DENIED -> FfiConsentState.DENIED
                UNKNOWN -> FfiConsentState.UNKNOWN
            }

        fun fromFfiConsentState(option: FfiConsentState): ConsentState =
            when (option) {
                FfiConsentState.ALLOWED -> ALLOWED
                FfiConsentState.DENIED -> DENIED
                FfiConsentState.UNKNOWN -> UNKNOWN
            }
    }
}

enum class EntryType {
    CONVERSATION_ID,
    INBOX_ID,
    ;

    companion object {
        fun toFfiConsentEntityType(option: EntryType): FfiConsentEntityType =
            when (option) {
                CONVERSATION_ID -> FfiConsentEntityType.CONVERSATION_ID
                INBOX_ID -> FfiConsentEntityType.INBOX_ID
            }

        fun fromFfiConsentEntityType(option: FfiConsentEntityType): EntryType =
            when (option) {
                FfiConsentEntityType.CONVERSATION_ID -> CONVERSATION_ID
                FfiConsentEntityType.INBOX_ID -> INBOX_ID
            }
    }
}

enum class PreferenceType {
    HMAC_KEYS,
}

data class ConsentRecord(
    val value: String,
    val entryType: EntryType,
    val consentType: ConsentState,
) {
    companion object {
        fun conversationId(
            groupId: String,
            type: ConsentState = ConsentState.UNKNOWN,
        ): ConsentRecord = ConsentRecord(groupId, EntryType.CONVERSATION_ID, type)

        fun inboxId(
            inboxId: InboxId,
            type: ConsentState = ConsentState.UNKNOWN,
        ): ConsentRecord = ConsentRecord(inboxId, EntryType.INBOX_ID, type)
    }

    val key: String
        get() = "${entryType.name}-$value"
}

data class PrivatePreferences(
    var client: Client,
    private val ffiClient: FfiXmtpClient,
) {
    suspend fun sync() {
        ffiClient.syncPreferences()
    }

    @Deprecated(message = "Use method `sync()` instead", replaceWith = ReplaceWith("sync()"))
    suspend fun syncConsent() {
        ffiClient.sendSyncRequest()
    }

    suspend fun streamPreferenceUpdates(onClose: (() -> Unit)? = null): Flow<PreferenceType> =
        callbackFlow {
            val preferenceCallback =
                object : FfiPreferenceCallback {
                    override fun onPreferenceUpdate(preference: List<FfiPreferenceUpdate>) {
                        preference.iterator().forEach {
                            when (it) {
                                is FfiPreferenceUpdate.Hmac -> trySend(PreferenceType.HMAC_KEYS)
                            }
                        }
                    }

                    override fun onError(error: FfiException) {
                        Log.e("XMTP preference update stream", error.toString())
                    }

                    override fun onClose() {
                        onClose?.invoke()
                        close()
                    }
                }

            val stream = ffiClient.conversations().streamPreferences(preferenceCallback)

            awaitClose { stream.end() }
        }

    suspend fun streamConsent(onClose: (() -> Unit)? = null): Flow<ConsentRecord> =
        callbackFlow {
            val consentCallback =
                object : FfiConsentCallback {
                    override fun onConsentUpdate(consent: List<FfiConsent>) {
                        consent.iterator().forEach {
                            trySend(it.fromFfiConsent())
                        }
                    }

                    override fun onError(error: FfiException) {
                        Log.e("XMTP consent stream", error.toString())
                    }

                    override fun onClose() {
                        onClose?.invoke()
                        close()
                    }
                }

            val stream = ffiClient.conversations().streamConsent(consentCallback)

            awaitClose { stream.end() }
        }

    suspend fun setConsentState(entries: List<ConsentRecord>) {
        ffiClient.setConsentStates(entries.map { it.toFfiConsent() })
    }

    private fun ConsentRecord.toFfiConsent(): FfiConsent =
        FfiConsent(
            EntryType.toFfiConsentEntityType(entryType),
            ConsentState.toFfiConsentState(consentType),
            value,
        )

    private fun FfiConsent.fromFfiConsent(): ConsentRecord =
        ConsentRecord(
            entity,
            EntryType.fromFfiConsentEntityType(entityType),
            ConsentState.fromFfiConsentState(state),
        )

    suspend fun conversationState(groupId: String): ConsentState =
        ConsentState.fromFfiConsentState(
            ffiClient.getConsentState(
                FfiConsentEntityType.CONVERSATION_ID,
                groupId,
            ),
        )

    suspend fun inboxIdState(inboxId: InboxId): ConsentState =
        ConsentState.fromFfiConsentState(
            ffiClient.getConsentState(
                FfiConsentEntityType.INBOX_ID,
                inboxId,
            ),
        )
}
